### PR TITLE
fix(eks): can't deploy with Bottlerocket amiType 

### PR DESCRIPTION
**Pull request by pahud**
As `Nodegroup` now supports different AMI types including Bottlerocket for both x86_64 and ARM_64, we cannot determine correct amiType simply from the `instanceTypes` property(#17641 ). However, when `instanceTypes` are provided, we still can check: 

1. if instance types of different CPU architectures are mxied and throw an error
2. if the provided `amyType` compatible with the instanceTypes

If user opt in Bottlerocket or any other AMI types other than AL2, users have to specify the `amiType` explicitly. If it's unspecified, we will use AL2 implicitly to avoid breaking changes, which is the default behavior previously.

The only case `amiType` has to be undefined is that when custom AMI is defined in the launch template. As we can't check this case, users have to explicitly leave it undefined. We add a notice in the property doc string for this.

Related to https://github.com/aws/aws-cdk/pull/12441
Fixes: #17641 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by pahud** [comment]
Hey @neilkuan I'll need your initial review. Let's discuss this offline.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 1]
```diff
@@ -357,15 +358,21 @@ export class Nodegroup extends Resource implements INodegroup {
...
     if (instanceTypes && instanceTypes.length > 0) {
-      // if the user explicitly configured instance types, we can calculate the expected ami type.
-      expectedAmiType = getAmiType(instanceTypes);
-
-      // if the user explicitly configured an ami type, make sure its the same as the expected one.
-      if (props.amiType && props.amiType !== expectedAmiType) {
-        throw new Error(`The specified AMI does not match the instance types architecture, either specify ${expectedAmiType} or dont specify any`);
+      /**
+       * if the user explicitly configured instance types, we can't caculate the expected ami type as we support
+       * Amazon Linux 2 and Bottlerocket now. However we can check:
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:366]: Could we do the opposite? Take the AMI type as a starting point (if it's provided) and figure out which instance types are valid for it?

Comment (neilkuan) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:366]: sounds like good, wdyt @pahud.

Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:366]: @otaviomacedo Yes this is exactly what I am working on.

I introduced `possibleAmiTypes` in this PR and calculate from the provided instance types.

Let's say if you provide all x86 instance types the `possibleAmiTypes` would only be ` AL2_X86_64` and `BOTTLEROCKET_X86_64`. In this case if `amiType` is provided we can check if it's included in the `possibleAmiTypes`.

**Comment by neilkuan** [comment]
for testing I know about if not specific `amitype` default will be `AL2_x86_64`.

so if I define `instanceTypes` to `['t4g.medium']`, will get error message   from cloudformation.
> [t4g.medium] is not a valid instance type for requested amiType AL2_x86_64 (Service: AmazonEKS; Status Code: 400; Error Code: InvalidParameterException; Request ID: 51f86ddf-0218-42c4-86d7-b02ffbe18f76; Proxy: null)

telegram-cloud-photo-size-5-6143385765976059443-y

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 2]
```diff
@@ -443,24 +456,40 @@ export class Nodegroup extends Resource implements INodegroup {
...
 // this function examines the CPU architecture of every instance type and determines
-// what ami type is compatible for all of them. it either throws or produces a single value because
+// what ami type is compatible for all of them. it either throws or produces an array of possible instance types because
 // instance types of different CPU architectures are not supported.
 function getAmiType(instanceTypes: InstanceType[]) {
-  const amiTypes = new Set(instanceTypes.map(i => getAmiTypeForInstanceType(i)));
-  if (amiTypes.size == 0) { // protective code, the current implementation will never result in this.
+  // const amiTypes = new Set(instanceTypes.map(i => getAmiTypeForInstanceType(i)));
+  const amiTypes = new Set<NodegroupAmiType>();
+  for (let t of instanceTypes) {
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:477]: can use `const` here.

```suggestion
  for (const instanceType of instanceTypes) {
```

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 3]
```diff
@@ -443,24 +456,40 @@ export class Nodegroup extends Resource implements INodegroup {
   }
 }
 
+const arm64AmiTypes: NodegroupAmiType[] = [NodegroupAmiType.AL2_ARM_64, NodegroupAmiType.BOTTLEROCKET_ARM_64];
+const x8664AmiTypes: NodegroupAmiType[] = [NodegroupAmiType.AL2_X86_64, NodegroupAmiType.BOTTLEROCKET_X86_64];
+const gpuAmiTypes: NodegroupAmiType[] = [NodegroupAmiType.AL2_X86_64_GPU];
+
 function getAmiTypeForInstanceType(instanceType: InstanceType) {
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:463]: 1. suggest to rename the function to `getAmiTypesForInstanceType` or `getPossibleAmiTypesForInstanceType`.
2. can specify return type on function declaration

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 4]
```diff
@@ -443,24 +456,40 @@ export class Nodegroup extends Resource implements INodegroup {
...
+      INSTANCE_TYPES.gpu.includes(instanceType.toString().substring(0, 2)) ? gpuAmiTypes :
+        INSTANCE_TYPES.inferentia.includes(instanceType.toString().substring(0, 4)) ? gpuAmiTypes :
+          x8664AmiTypes;
 }
 
 // this function examines the CPU architecture of every instance type and determines
-// what ami type is compatible for all of them. it either throws or produces a single value because
+// what ami type is compatible for all of them. it either throws or produces an array of possible instance types because
 // instance types of different CPU architectures are not supported.
 function getAmiType(instanceTypes: InstanceType[]) {
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:474]: 1. suggest to rename the function to `getAmiTypes` or `getPossibleAmiTypes`
2. can specify return type on function declaration

**Review by kirintwn — COMMENTED** [review]
Some NITS. other than that, looks good.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 3 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts]: Thanks! Just fixed.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 4 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts]: all addressed

**Comment by pahud** [comment]
> Some NITS. Other than that, looks good. Just need to be cautious that we need to calculate the recommended `amiType` corresponding to the `instanceTypes` when the user doesn't specify `amiType` or ami in `launchTemplate`. Also need to care the backward compatibility when calculating the recommended `amiType` for user.
> 
> Maybe add something like this in the `if (instanceTypes && instanceTypes.length > 0)` block:
> 
> ```ts
> if (!props.amiType && !props.launchTemplateSpec) {
>   // calculate the recommended `amiType` corresponding to the `instanceTypes`
> }
> ```
> 
> or when constructing the `CfnNodegroup`, use:
> 
> ```ts
> const resource = new CfnNodegroup(this, 'Resource', {
>   ...
>   amiType: props.launchTemplateSpec ? props.amiType : (props.amiType ?? possibleAmiTypes[0]),
>   ...
> });
> ```

Let's dive a little bit deeper here. We have some optional properties in this construct:

1.  `props.instanceTypes` is optional
2. `props.launchTemplateSpec` is optional
3. `props.amiType` is optional

When `props.instanceTypes` is specified, there might be some cases:

1. `amiType` undefined with `launchTemplateSpec` undefined
Check:
   - we just have to ensure the instanceTypes provided are compatible with the default amiType, i.e. `AL2_X86_64`.
2. `amiType` undefined with `launchTemplateSpec` defined
Yes it's possible to have `launchTemplateSpec` defined while the instance types are also defined in `props.instanceTypes` property.
Check:
   - Same as above.
3. `amiType` defined with `launchTemplateSpec` undefined
  - we check if props.amiType is included in `possibleAmiTypes` we generate from `props.instanceTypes`.
4. `amiType` defined with `launchTemplateSpec` defined
  - we still have to check if `props.amiType` is included in `possibleAmiTypes` we generate from `props.instanceTypes`. As we can't see if AMI is also defined in the launch template, it's user responsibility to leave `props.amiType` undefiled if AMI is also specified in the launch template.
  
That being said, it doesn't make difference where props.launchTemplateSpec is defined or not. The validation check is literally same.

**Comment by pahud** [comment]
Hi @neilkuan @kirintwn I am ready.  Can you take another look?

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 5]
```diff
@@ -357,15 +358,21 @@ export class Nodegroup extends Resource implements INodegroup {
...
+       * Amazon Linux 2 and Bottlerocket now. However we can check:
+       *
+       * 1. instance types of different CPU architectures are not mixed(e.g. X86 with ARM).
+       * 2. user-specified amiType should be included in `possibleAmiTypes`.
+       */
+      possibleAmiTypes = getPossibleAmiTypes(instanceTypes);
+
+      // if the user explicitly configured an ami type, make sure it's included in the possibleAmiTypes
+      if (props.amiType && !possibleAmiTypes.includes(props.amiType)) {
+        throw new Error(`The specified AMI does not match the instance types architecture, either specify one of ${possibleAmiTypes} or dont specify any`);
```
Comment (neilkuan) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:375]:
```suggestion
        throw new Error(`The specified AMI does not match the instance types architecture, either specify one of ${possibleAmiTypes} or don't specify any`);
```

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 6]
```diff
@@ -157,9 +157,10 @@ export interface NodegroupOptions {
    */
   readonly subnets?: SubnetSelection;
   /**
-   * The AMI type for your node group.
+   * The AMI type for your node group. If you explicitly specify the launchTemplate with custom AMI, do not specify this property, or
+   * the node group deployment will fail. In other cases, you will need to specify correct amiType for the nodegroup.
    *
-   * @default - auto-determined from the instanceTypes property.
+   * @default - undefined.
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:163]: This should be something like:
`auto-determined from the instanceTypes property when launchTemplateSpec property is not specified`

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 7]
```diff
@@ -443,24 +456,48 @@ export class Nodegroup extends Resource implements INodegroup {
...
 }
 
 // this function examines the CPU architecture of every instance type and determines
-// what ami type is compatible for all of them. it either throws or produces a single value because
+// what ami type is compatible for all of them. it either throws or produces an array of possible instance types because
 // instance types of different CPU architectures are not supported.
-function getAmiType(instanceTypes: InstanceType[]) {
-  const amiTypes = new Set(instanceTypes.map(i => getAmiTypeForInstanceType(i)));
-  if (amiTypes.size == 0) { // protective code, the current implementation will never result in this.
+function getPossibleAmiTypes(instanceTypes: InstanceType[]): NodegroupAmiType[] {
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:483]: Can add JSDoc docstring for the function like:
```typescript
/**
 * This function returns an array of "possible AMI types" from the given instance types array
 * @param instanceTypes The instance types array
 * @returns NodegroupAmiType[]
 */
```

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 8]
```diff
@@ -443,24 +456,48 @@ export class Nodegroup extends Resource implements INodegroup {
...
+  const amiTypes = new Set<NodegroupAmiType>();
+  for (const t of instanceTypes) {
+    getPossibleAmiTypesForInstanceType(t).forEach(x => amiTypes.add(x));
+  }
+  if (new Set(amiTypes).size == 0) { // protective code, the current implementation will never result in this.
```
Comment (kirintwn) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:488]: 1. use `===`
2. don't think we need to construct a new set here

```suggestion
  if (amiTypes.size === 0) { // protective code, the current implementation will never result in this.
```

**Review by neilkuan — COMMENTED** [review]
1. Can you also add testing for got this error message `instanceTypes of different CPU architectures is not allowed` ?!
2. Can you add the `c7g` in to [instance-types.ts](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-eks/lib/instance-types.ts) for  graviton3?! 
other LGTM.

**Review by kirintwn — COMMENTED** [review]
Please add some happy test combinations:
```
instanceTypes: specified
amiType: specified
launchTemplateSpec: unspecified
```
Especially cases in the reproduction steps in the issue #17641, since we are fixing that issue.

**Comment by pahud** [comment]
> 1. Can you also add testing for got this error message `instanceTypes of different CPU architectures is not allowed` ?!
>    ex:
> 
> ```ts
>  this.cluster.addNodegroupCapacity('BottlerocketNG1', {
>       instanceTypes: [new ec2.InstanceType('t4g.large'), new ec2.InstanceType('t2.large')],
>     });
> ```

The test is already there.
https://github.com/pahud/aws-cdk/blob/d3acc17b3fa9e9c928d05ff9e85f4f1fb508e053/packages/%40aws-cdk/aws-eks/test/nodegroup.test.ts#L728-L745

> 
> 2. Can you add the `c7g` in to [instance-types.ts](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-eks/lib/instance-types.ts) for  graviton3?!

Sure.
>    other LGTM.

File: packages/@aws-cdk/aws-eks/test/nodegroup.test.ts [hunk 9]
```diff
@@ -230,9 +230,284 @@ describe('node group', () => {
...
+    cluster.addNodegroupCapacity('bottlerocket', {
+      instanceTypes: [new ec2.InstanceType('m5a.xlarge')],
+      amiType: NodegroupAmiType.BOTTLEROCKET_X86_64,
+    });
+
+    // THEN
+    expect(stack).toHaveResourceLike('AWS::EKS::Nodegroup', {
+      AmiType: 'BOTTLEROCKET_x86_64',
+    });
+  });
```
Comment (pahud) [inline packages/@aws-cdk/aws-eks/test/nodegroup.test.ts:486]: A special unit test for the reported issue. FYR @kirintwn

File: packages/@aws-cdk/aws-eks/test/nodegroup.test.ts [hunk 10]
```diff
@@ -230,9 +230,284 @@ describe('node group', () => {
...
+      instanceTypes: [new ec2.InstanceType('c7g.xlarge')],
+      amiType: NodegroupAmiType.BOTTLEROCKET_ARM_64,
+    });
+
+    // THEN
+    expect(stack).toHaveResourceLike('AWS::EKS::Nodegroup', {
+      AmiType: 'BOTTLEROCKET_ARM_64',
+    });
+  });
+
```
Comment (pahud) [inline packages/@aws-cdk/aws-eks/test/nodegroup.test.ts:510]: This test is for Gravition 3. @neilkuan

File: packages/@aws-cdk/aws-eks/test/nodegroup.test.ts (hunk 9 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/test/nodegroup.test.ts:486]: m5a.xlarge with BOTTLEROCKET_X86_64 successfully deployed.

image

image

**Comment by pahud** [comment]
@neilkuan @kirintwn all addressed. Please take another look. Thanks.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 11]
```diff
@@ -443,24 +456,52 @@ export class Nodegroup extends Resource implements INodegroup {
...
+
+/**
+ * This function returns an array of "possible AMI types" from the given instance type
+ * @param instanceType The instance type
+ * @returns NodegroupAmiType[]
+ */
+function getPossibleAmiTypesForInstanceType(instanceType: InstanceType): NodegroupAmiType[] {
+  return INSTANCE_TYPES.graviton2.includes(instanceType.toString().substring(0, 3)) ? arm64AmiTypes :
+    INSTANCE_TYPES.graviton.includes(instanceType.toString().substring(0, 2)) ? arm64AmiTypes :
+      INSTANCE_TYPES.gpu.includes(instanceType.toString().substring(0, 2)) ? gpuAmiTypes :
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:475]: The `ec2.InstanceType` class has an `architecture` method, that already has some of this logic, except that it's not aware of GPU. Maybe we should add the missing bit there and reuse it.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts [hunk 12]
```diff
@@ -443,24 +456,52 @@ export class Nodegroup extends Resource implements INodegroup {
...
+function getPossibleAmiTypes(instanceTypes: InstanceType[]): NodegroupAmiType[] {
+  const amiTypes = new Set<NodegroupAmiType>();
+  for (const t of instanceTypes) {
+    getPossibleAmiTypesForInstanceType(t).forEach(x => amiTypes.add(x));
+  }
+  if (amiTypes.size === 0) { // protective code, the current implementation will never result in this.
     throw new Error(`Cannot determine any ami type comptaible with instance types: ${instanceTypes.map(i => i.toString).join(',')}`);
   }
-  if (amiTypes.size > 1) {
+  let cpuArchTypes: number = 0;
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:495]: Let me suggest a slightly different way, that IMO is easier to understand:
1. From the list of instance types, compute the set of architectures (see my other comment).
2. If the set has 0 or more than one element, throw an error.
3. At this point, there is only one architecture. Return the list of types that correspond to it (combine `arm64AmiTypes`, `x8664AmiTypes` and `gpuAmiTypes` into a single `Map<InstanceArchitecture, NodegroupAmiType[]>`, for example).

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 11 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts:475]: Agree. Let me give it a try.

Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts]: fixed.

File: packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts (hunk 12 above)
Comment (pahud) [reply packages/@aws-cdk/aws-eks/lib/managed-nodegroup.ts]: fixed

**Comment by pahud** [comment]
Hi @otaviomacedo 

I have revised the implementation based on your suggested changes. I tend not to modify aws-ec2 in this PR so I have a tiny `isGpuInstanceType()` helper function.

Let me know if you have any other thoughts.

copy @neilkuan @kirintwn

**Comment by pahud** [comment]
> > I have revised the implementation based on your suggested changes. I tend to not modify aws-ec2 in this PR so I have a tiny `isGpuInstanceType()` helper function.
> 
> Yeah, I see how changing `InstanceType` would break existing code. But still, there is an opportunity for simplification here. Something like:
> 
> ```ts
> type MyArchitecture = InstanceArchitecture | 'GPU';
> 
> function getPossibleAmiTypes(instanceTypes: InstanceType[]): NodegroupAmiType[] {
>   function typeToArch(instanceType: InstanceType): MyArchitecture {
>     return isGpuInstanceType(instanceType) ? 'GPU' : instanceType.architecture;
>   }
> 
>   const archAmiMap = new Map<MyArchitecture, NodegroupAmiType[]>([
>     [InstanceArchitecture.ARM_64, arm64AmiTypes],
>     [InstanceArchitecture.X86_64, x8664AmiTypes],
>     ['GPU', gpuAmiTypes],
>   ]);
> 
>   const architectures: Set<MyArchitecture> = new Set(instanceTypes.flatMap(typeToArch));
> 
>   if (architectures.size === 0) {
>     throw new Error(`Cannot determine any ami type comptaible with instance types: ${instanceTypes.map(i => i.toString).join(',')}`);
>   }
> 
>   if (architectures.size > 1) {
>     throw new Error('instanceTypes of different architectures is not allowed');
>   }
> 
>   return archAmiMap.get(Array.from(architectures)[0])!;
> }
> ```

@otaviomacedo  This is interesting! Let me try it.

**Comment by pahud** [comment]
@otaviomacedo all fixed. Please take a look.

**Comment by pahud** [comment]
@kirintwn all resolved :-)

**Comment by kirintwn** [comment]
@pahud:
> And is it okay to let the docstring separated from the function with this type declaration?
Will it breaks auto-generation mechanism of documentation?

https://github.com/aws/aws-cdk/blob/cd5f79cfa2d822f7716107e96abf9e91b22efe94/packages/%40aws-cdk/aws-eks/lib/managed-nodegroup.ts#L481-L489

can you confirm this?

**Comment by pahud** [comment]
> can you confirm this? thanks!

@kirintwn nice catch! all fixed!

Maybe another look? copy @otaviomacedo

File: .devcontainer.json [hunk 13]
```diff
@@ -1,8 +1,8 @@
 {
     "name": "Dev Container Definition - AWS CDK",
-    "image": "jsii/superchain",
+    "image": "mcr.microsoft.com/vscode/devcontainers/universal",
```
Comment (kirintwn) [inline .devcontainer.json:3]: is this change intentional? Seems not to be related to the PR's scope

Comment (pahud) [reply .devcontainer.json:3]: oh...my bad. I should not submit this.

**Review by kirintwn — APPROVED** [review]
LGTM

**Review by otaviomacedo — APPROVED** [review]
